src/Network/Wai/Middleware/Auth/OAuth2/Github.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+module Network.Wai.Middleware.Auth.OAuth2.Github
+    ( Github(..)
+    , githubParser
+    , mkGithubProvider
+    ) where
+import           Control.Exception.Safe               (catchAny)
+import           Control.Monad                        (guard)
+import           Data.Aeson
+import qualified Data.ByteString                      as S
+import           Data.Monoid                          ((<>))
+import qualified Data.Text                            as T
+import           Data.Text.Encoding                   (encodeUtf8)
+import           Network.HTTP.Simple
+import           Network.HTTP.Types
+import           Network.Wai.Auth.Tools               (getValidEmail)
+import           Network.Wai.Middleware.Auth.OAuth2
+import           Network.Wai.Middleware.Auth.Provider
+
+
+mkGithubProvider
+  :: T.Text
+  -> T.Text
+  -> T.Text
+  -> [T.Text]
+  -> Github
+mkGithubProvider appName clientId clientSecret emailWhiteList =
+  Github
+    appName
+    "https://api.github.com/user/emails"
+    emailWhiteList
+    OAuth2
+    { oa2ClientId = clientId
+    , oa2ClientSecret = clientSecret
+    , oa2AuthorizeEndpoint = "https://github.com/login/oauth/authorize"
+    , oa2AccessTokenEndpoint = "https://github.com/login/oauth/access_token"
+    , oa2Scope = Just ["user:email"]
+    , oa2ProviderInfo =
+        ProviderInfo
+        { providerTitle = "GitHub"
+        , providerLogoUrl =
+            "https://assets-cdn.github.com/images/modules/logos_page/Octocat.png"
+        , providerDescr =
+            "Use your GitHub account to access this page."
+        }
+    }
+
+githubParser :: ProviderParser
+githubParser = mkProviderParser (undefined :: Github)
+
+
+data Github = Github
+  { githubAppName          :: T.Text
+  , githubAPIEmailEndpoint :: T.Text
+  , githubEmailWhitelist   :: [T.Text]
+  , githubOAuth2           :: OAuth2
+  }
+
+instance FromJSON Github where
+
+  parseJSON = withObject "Github Provider Object" $ \ obj -> do
+    appName <- obj .: "app_name"
+    clientId <- obj .: "client_id"
+    clientSecret <- obj .: "client_secret"
+    emailWhiteList <- obj .:? "email_white_list" .!= []
```

Comment:
I would go for strict defaults too and empty list would not match any emails. But manually setting a whitelist to pattern `[".*"]`, would match all.